### ISSUE TITLE
avoid double forward pass for train metrics

### ISSUE THREAD

**Issue by A-Jacobson** [mosaicml/composer#467]
Our current training loop requires an extra forward pass to compute training metrics which creates a considerable slowdown. Since training metrics are pretty useful, this slowdown could be a dealbreaker for some. It also seems completely avoidable, I see two options:

current loop..ish
```python

def fit():
    while timer < max_duration:
        model.train()  # set model to train mode
        for batch in train_dataloader:

            if self.compute_training_metrics:
                model.eval()
                with torch.no_grad():
                    outputs, targets = model.validate(batch)
                    train_metrics.update(outputs, targets)
                model.train()

            optimizer.zero_grad()
            outputs = model.forward(batch)
            loss = model.loss(outputs, batch)
            loss.backward()
            scheduler.step()

        ## Eval
        model.eval()
        eval_metrics = model.metrics(train=False)  # get eval metrics
        with torch.no_grad():
            for batch in eval_dataloader:
                outputs, targets = model.validate(batch)
                eval_metrics.update(outputs, targets)
            eval_metrics.compute()  # compute and sync metrics across devices
```
###  Option 1 :
 re-use outputs from forward and access targets from `state.batch`. 

upside:
-  solves the problem
-  no changes to the model class

downside:
- we need to know how to unpack the batch =(

```python
  for batch in train_dataloader:
      optimizer.zero_grad()
      outputs = model.forward(batch)
      loss = model.loss(outputs, batch)
      if self.compute_training_metrics:
          train_metrics.update(outputs.detach(), state.batch[1]) # tuple batch
      loss.backward()
      scheduler.step()
 ```
 ### Option 2:
 return `outputs` and `targets` from `model.forward`
upside:
-  solves the problem
- user specifies how to unpack the batch
- potential to unify `forward` and `validate` by having `validate` call forward by default and allowing users to override if their validation loop requires special logic such as beamsearch decoding for nlp. 

downside:
- targets return from `forward` is sort of strange, but we cant use `forward` for prediction for our current setup anyway as it expects a full batch (with targets). maybe we rename this method as there's already an expectation (though not convention) in pytorch that `forward` takes model input and returns output?
- `validate` overrides would either be ignored in the training loop, or require two forward steps yet again.

```python
  for batch in train_dataloader:
      optimizer.zero_grad()
      outputs, targets = model.forward(batch)
      loss = model.loss(outputs, batch)
      if self.compute_training_metrics:
          train_metrics.update(outputs.detach(), targets)
      loss.backward()
      scheduler.step()
 ```
 
 current  `ComposerClassifier`'s methods, note they're pretty much the same. The only model we have that currently has a special `validation` method is `Unet` to the best of my knowledge.
 ```python
    def forward(self, batch: BatchPair) -> Tensor:
        x, _ = batch
        logits = self.module(x)
        return logits

    def validate(self, batch: BatchPair) -> Tuple[Any, Any]:
        assert self.training is False, "For validation, model must be in eval mode"
        _, targets = batch
        logits = self.forward(batch)
        return logits, targets
```

proposed, though maybe validate is set to call forward on the base class.
```python
    def forward(self, batch: BatchPair) -> Tensor:
        inputs, targets = batch
        outputs = self.module(inputs)
        return outputs

    def validate(self, batch: BatchPair) -> Tuple[Any, Any]:
        return self.forward(batch)
```
@hanlint, @Landanjs , @anisehsani

**Comment by Landanjs** [comment]
I like Option 2 since I think the name `forward()` is already confusing for the utility of the function (i.e. split's the batch for forward pass and splits the model output for `loss()`). I'll have to think which name makes more sense to me...

For Unet, I think the unique `validate()` function is to handle 3D images from the eval dataloader. Maybe this processing can be moved to eval dataloader or performed as preprocessing? We can chat more.

Higher-level, naive question: what do you think is the utility of the train metrics calculated with `model.train()` and train data preprocessing? The metrics could become very different with certain algorithms e.g. stochastic depth or cutmix.

I understand we should still have the functionality since people expect it. But should we make it clear that the metrics may be very different than providing the training dataset as another evaluation dataset (with eval preprocessing)? Although there will be a slowdown, the measurement should be more useful for research.

**Comment by A-Jacobson** [comment]
I'm not sure yet about the first two. Definitely going to be ongoing discussions. For example, with unet in particular it's probably better to move the preprocessing to the loader. But in general I don't think we should expect people to modify their datasets.

Regarding training metrics. I think it's useful to compare them to validation metrics during training, even if they are very noisy. Obviously augmentation and stuff like mixup are going to limit the usefulness of these but most people still do log them by default. I also believe it's not unreasonable to assume people know training metrics are noisy since they're normally a moving average of a model that's still being trained. If we wanted truly accurate metrics we would have to do a full pass through the train set at the end of each epoch just as we do with the validation set. There's probably some value in doing this if you really want to make assumptions about overfitting or how well matched your train and validation sets are but I'm not sure how common this is.

**Comment by hanlint** [comment]
In most cases, where the output of `forward` can be consumed by `metrics`, this is correct. However, our API only enforces `forward -> loss` and `validate -> metrics`, so this seems like a feature that would break that contract.

**Comment by dblalock** [comment]
Brainstorming: what if we had an optional `prep_for_metrics` method? We wouldn't enforce that `forward -> metrics` worked, but we would enforce that `forward, batch -> prep_for_metrics -> metrics` worked. Users doing simple things will never need to touch this method, but would allow arbitrary flexibility.

Might also let us get rid of validate? (low confidence here because I don't really understand why it exists right now)


Also, fwiw, I see avoidance of an extra forward pass for training metrics as a requirement for our API; i.e., if the API doesn't allow it, the API needs to change.

**Comment by A-Jacobson** [comment]
> Brainstorming: what if we had an optional `prep_for_metrics` method? We wouldn't enforce that `forward -> metrics` worked, but we would enforce that `forward, batch -> prep_for_metrics -> metrics` worked. Users doing simple things will never need to touch this method, but it would allow arbitrary flexibility for those who need it.
> 
> Might also let us get rid of validate? (low confidence here because I don't really understand why it exists right now)
> 
> Also, fwiw, I see avoidance of an extra forward pass for training metrics as a requirement for our API; i.e., if the API doesn't allow it, the API needs to change.

Have been thinking of something along these lines ,but perhaps making `metrics` into more of a `run_metrics` type method that actually explicitly takes the outputs of `forward` and computes metrics on them. This way the user can just manually manipulate what they need and they know exactly what's being computed rather than us hiding stuff from them in the training loop. 

I'd like to avoid breaking up the linear, explicit training loops people are used to with a bunch of callbacks and callable if it all possible. 

I haven't looked at all the implications but it would be something like this:
```python
def metrics(outputs, batch, train=False):
   inputs, targets = batch
   if train:
    accuracy = self.train_acc(outputs, targets)
    loss = self.train_loss(outputs, targets)
    return accuracy, loss
  accuracy = self.val _acc(outputs, targets)
  loss = self.val_metric(outputs, targets)
  return accuracy, loss

```

**Comment by mvpatel2000** [comment]
This is now implemented :)
